Get multiple properties

rickardoberg opened the pull request:
This adds support for retrieving multiple properties at a time from nodes/rels, and updates various client code that previously used getPropertyKeys and looping to use this new method instead.

rickardoberg committed:
Add getProperties(names) and getAllProperties to PropertyContainer

rickardoberg committed:
Update client code, tests, and mocks to use the new get-all-properties methods

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java:
```diff
@@ -377,6 +382,85 @@ public Object getProperty( String key, Object defaultValue )
     }
 
     @Override
+    public Map<String, Object> getProperties( String... keys )
```
We can add a check `keys == null || keys.length == 0` and do a quick return of `Collections.emptyMap()`

burqen replied:
`keys` should be `names`

davidegrohmann replied:
I think we should throw if `keys == null` or if `keys` contains `null` values.
We can return `emptyMap()` when `keys.length == 0`.

rickardoberg replied:
@burqen no, we use the word "keys" in the API (e.g. getPropertyKeys), so the correct word is indeed keys

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java:
```diff
@@ -377,6 +382,85 @@ public Object getProperty( String key, Object defaultValue )
+                            int propertyKey = propertyKeys[i];
+                            if ( propertyKey == propertyCursor.get().propertyKeyId() )
+                            {
+                                properties.put( keys[i], propertyCursor.get().value() );
```
how about `break;` when we have found the matching key?

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java:
```diff
@@ -236,6 +241,85 @@ public RelationshipType getType()
+                            new EntityNotFoundException( EntityType.RELATIONSHIP, getId() ) );
+                }
+
+                try ( Cursor<PropertyItem> propertyCursor = relationship.get().properties() )
```
Can we create a utility-class that contains a utility method for this code that is duplicated between NodeProxy and RelationshipProxy?

davidegrohmann replied:
+1

rickardoberg replied:
I'll put it in an abstract super class for NodeProxy/RelationshipProxy

lutovich commented on community/kernel/src/test/java/org/neo4j/kernel/impl/core/NodeTest.java:
```diff
@@ -326,9 +327,14 @@ public void testNodeGetProperties()
+        assertTrue( properties.get( key1 ).equals( int1 ) );
+        assertTrue( properties.get( key2 ).equals( int2 ) );
+        assertTrue( properties.get( key3 ).equals( string ) );
+        properties = node1.getProperties( key1, key2 );
```
Missing tests for negative cases where given array of keys is empty or null

davidegrohmann replied:
I would also like a test when the array contains `null` values.

lutovich commented on community/kernel/src/test/java/org/neo4j/kernel/impl/core/TestRelationship.java:
```diff
@@ -557,6 +559,16 @@ public void testRelGetProperties()
+        assertTrue( properties.get( key1 ).equals( int1 ) );
+        assertTrue( properties.get( key2 ).equals( int2 ) );
+        assertTrue( properties.get( key3 ).equals( string ) );
+        properties = rel1.getProperties( key1, key2 );
```
Missing tests for negative cases where given array of keys is empty or null

davidegrohmann replied:
I would also like a test when the array contains `null` values.

lutovich commented on community/kernel/src/main/java/org/neo4j/graphdb/PropertyContainer.java:
```diff
@@ -126,4 +128,19 @@
+     * Returns specified existing properties.
+     *
+     * @param keys the property keys to return
+     * @return specified properties on this property container
```
We should ask users to not rely on the mutability of the returned map and document what will happen if empty/null array of names is given as a parameter

davidegrohmann replied:
I hope we make sure that the returned map is immutable.

What about saying that we'll throw some exception (NPE or IAE) if we get `null` keys in the array?

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java:
```diff
@@ -377,6 +382,85 @@ public Object getProperty( String key, Object defaultValue )
+                            if ( propertyKey == propertyCursor.get().propertyKeyId() )
+                            {
+                                properties.put( keys[i], propertyCursor.get().value() );
+                            }
```
Would it be more expensive to have a more dynamic structure for `propertyKeys` and removing the keys that we already found to avoid looping over them again?

rickardoberg replied:
Yes, I'm trying to avoid generating unnecessary GC.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java:
```diff
@@ -377,6 +382,85 @@ public Object getProperty( String key, Object defaultValue )
+                    {
+                        String key = keys[i];
+                        propertyKeys[i] = statement.readOperations().propertyKeyGetForName( key );
+                    }
```
there is no reason to create an array here

rickardoberg replied:
So you think it would be better to do the name->int lookup on every check instead? Won't that impact performance a bit too much?

AlexanderYastrebov commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/GraphPropertiesProxy.java:
```diff
@@ -195,6 +198,45 @@ public Object removeProperty( String key )
+            for ( String name : names )
+            {
+                int propertyKeyId = readOperations.propertyKeyGetForName( name );
+                properties.put( name, readOperations.graphGetProperty( propertyKeyId ) );
```
Here you could end up with null values in map for absent properties. Is that ok?

thobe replied:
That might actually be desirable (since it means the set of keys is maintained), I wonder if we should make the other implementations do that as well, or if we should change this implementation, they should at least have the same behaviour in that regard (although this implementation is of course never used by end-user code).

This is something we should discuss IMO.

davidegrohmann replied:
+1 on the discussion.

rickardoberg replied:
I would prefer if only found properties were included, in all cases. That would match better the semantics that we already have in place, i.e. there cannot be null valued properties.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/GraphPropertiesProxy.java:
```diff
@@ -195,6 +198,45 @@ public Object removeProperty( String key )
+                int propertyKeyId = readOperations.propertyKeyGetForName( name );
+                properties.put( name, readOperations.graphGetProperty( propertyKeyId ) );
+            }
+            return properties;
```
Please return `Collections.unmodifiableMap( properties )` instead.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/GraphPropertiesProxy.java:
```diff
@@ -195,6 +198,45 @@ public Object removeProperty( String key )
+                properties.put( readOperations.propertyKeyGetName( propertyKeyId ),
+                        readOperations.graphGetProperty( propertyKeyId ) );
+            }
+            return properties;
```
Please return `Collections.unmodifiableMap( properties )` instead.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java:
```diff
@@ -377,6 +382,85 @@ public Object getProperty( String key, Object defaultValue )
+                        }
+                    }
+
+                    return properties;
```
Please return `Collections.unmodifiableMap( properties )` instead.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java:
```diff
@@ -377,6 +382,85 @@ public Object getProperty( String key, Object defaultValue )
+                        properties.put( name, propertyCursor.get().value() );
+                    }
+
+                    return properties;
```
Please return `Collections.unmodifiableMap( properties )` instead.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java:
```diff
@@ -377,6 +382,85 @@ public Object getProperty( String key, Object defaultValue )
+
+                try ( Cursor<PropertyItem> propertyCursor = node.get().properties() )
+                {
+                    Map<String, Object> properties = new HashMap<>( ((int) (keys.length * 1.3f)) );
```
Can we add a comment explaining the rationale behind `((int) (keys.length * 1.3f))`?

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java:
```diff
@@ -236,6 +241,85 @@ public RelationshipType getType()
+                    for ( int i = 0; i < keys.length; i++ )
+                    {
+                        String key = keys[i];
+                        propertyKeys[i] = statement.readOperations().propertyKeyGetForName( key );
```
no reason to create an array here

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java:
```diff
@@ -236,6 +241,85 @@ public RelationshipType getType()
+                        }
+                    }
+
+                    return properties;
```
Please return `Collections.unmodifiableMap( properties )` instead.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java:
```diff
@@ -236,6 +241,85 @@ public RelationshipType getType()
+                        properties.put( name, propertyCursor.get().value() );
+                    }
+
+                    return properties;
```
Please return `Collections.unmodifiableMap( properties )` instead.

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/test/mocking/Properties.java:
```diff
@@ -82,4 +82,9 @@ public Object answer( InvocationOnMock invocation ) throws Throwable
+
+    public SortedMap<String, Object> getProperties()
+    {
+        return properties;
```
Not super important, but why not `Collections.unmodifiableMap( properties )`?

rickardoberg replied:
Two reasons: first I want to minimize the impact of GC that these methods have. Second, if the client changes these maps nothing bad happens. So returning modifiable maps should be fine.

davidegrohmann commented on community/ndp/messaging-v1/src/main/java/org/neo4j/ndp/messaging/v1/infrastructure/ValueRelationship.java:
```diff
@@ -107,6 +108,26 @@ public Node getEndNode()
+        return props;
+    }
+
+    @Override
```
Can we avoid this duplication of code in `NodeValue` and `RelationshipValue`?

davidegrohmann commented on community/ndp/messaging-v1/src/main/java/org/neo4j/ndp/messaging/v1/infrastructure/ValueUnboundRelationship.java:
```diff
@@ -95,6 +93,26 @@ public long getId()
+        return props;
+    }
+
+    @Override
```
This code has been duplicate 3 times, please let's try to avoid so much duplication.

rickardoberg replied:
I'll introduce a common base class

davidegrohmann commented on community/shell/src/main/java/org/neo4j/shell/kernel/ReadOnlyGraphDatabaseProxy.java:
```diff
@@ -387,6 +387,18 @@ public Object getProperty( String key, Object defaultValue )
         @Override
+        public Map<String, Object> getProperties( String... names )
+        {
+            return actual.getProperties( names );
```
`Collections.unmodifiableMap( ... )`

davidegrohmann commented on community/shell/src/main/java/org/neo4j/shell/kernel/ReadOnlyGraphDatabaseProxy.java:
```diff
@@ -387,6 +387,18 @@ public Object getProperty( String key, Object defaultValue )
+        @Override
+        public Map<String, Object> getAllProperties()
+        {
+            return actual.getAllProperties();
```
`Collections.unmodifiableMap( ... )`

davidegrohmann commented on community/shell/src/main/java/org/neo4j/shell/kernel/ReadOnlyGraphDatabaseProxy.java:
```diff
@@ -536,6 +548,18 @@ public Object getProperty( String key, Object defaultValue )
         @Override
+        public Map<String, Object> getProperties( String... names )
+        {
+            return actual.getProperties( names );
```
`Collections.unmodifiableMap( ... )`

davidegrohmann commented on community/shell/src/main/java/org/neo4j/shell/kernel/ReadOnlyGraphDatabaseProxy.java:
```diff
@@ -536,6 +548,18 @@ public Object getProperty( String key, Object defaultValue )
+        @Override
+        public Map<String, Object> getAllProperties()
+        {
+            return actual.getAllProperties();
```
`Collections.unmodifiableMap( ... )`

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/core/TestRelationship.java:
```diff
@@ -557,6 +559,16 @@ public void testRelGetProperties()
+        assertTrue( properties.get( key1 ).equals( int1 ) );
+        assertTrue( properties.get( key2 ).equals( int2 ) );
+        assertFalse( properties.containsKey( key3 ) );
+
```
Can we please split this method in several ones with proper test-explanatory names?

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/core/NodeTest.java:
```diff
@@ -326,9 +327,14 @@ public void testNodeGetProperties()
+        properties = node1.getProperties( key1, key2 );
+        assertTrue( properties.get( key1 ).equals( int1 ) );
+        assertTrue( properties.get( key2 ).equals( int2 ) );
+        assertFalse( properties.containsKey( key3 ) );
```
Can we please split this method in several ones with proper test-explanatory names?

davidegrohmann commented:
Missing unit tests on `NodeValue`, `ValueRelationship`, `ValueUnboundRelationship`, `NodeProxy`, and `RelationshipProxy`?

rickardoberg committed:
Introduced base classes for shared node/rel code related to PropertyContainer implementation

Added more negative tests

Added Javadocs explaining what to expect

Added input parameter testing

lutovich commented:
I agree with @davidegrohmann on missing tests and needed refactoring in `TestRelationship` and `NodeTest`

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/PropertyContainerProxy.java:
```diff
@@ -0,0 +1,70 @@
+            String[] keys )
+    {
+        // Create a map which is slightly larger than the list of keys
+        Map<String, Object> properties = new HashMap<>( ((int) (keys.length * 1.3f)) );
```
can we extract creation of map that should not resize anymore to map util?

rickardoberg replied:
Why?

jexp replied:
Actually we should not use an HashMap but a fixed-size, efficient ArrayMap :)

rickardoberg replied:
No we shouldn't, since ArrayMap does not implement Map. And we're NOT going to expose ArrayMap from our API. Other suggestions?

MishaDemianenko replied:
because that sounds like handy things to do if you want to avoid useless resizing. 
But will agree with @jexp suggestion :)

MishaDemianenko replied:
Then falling back to my original comment, that sounds handy to have helper method that will create map with the size that you exactly need.

thobe replied:
Last time I checked `org.neo4j.kernel.impl.util.ArrayMap` was all but "efficient", and definitely not "fixed-size". We might want to write some new Map implementation for this purpose, but we definitely would not want to reuse `org.neo4j.kernel.impl.util.ArrayMap` here. I furthermore agree with @rickardoberg that the cost of writing our own Map implementation does not seem worth it for this use case. Even if the cost of this conversation is starting to creep up on being as high as writing our own Map implementation would be.

My vote is thus, let the bike shed remain [dark orchid](http://darkorchid.bikeshed.org/).
